File: solutions/JavaScript/merge_two_sorted_lists_recursive.js
class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

/**
 * Merge two sorted linked lists recursively.
 *
 * Time Complexity: O(m + n) where m and n are the lengths of list1 and list2
 * Space Complexity: O(m + n) due to the recursion call stack
 *
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
function mergeTwoSortedListsRecursive(list1, list2) {
    // Base cases: if one list is empty, return the other
    if (!list1) {
        return list2;
    }
    if (!list2) {
        return list1;
    }

    // Compare the values and recursively merge
    if (list1.val <= list2.val) {
        list1.next = mergeTwoSortedListsRecursive(list1.next, list2);
        return list1;
    } else {
        list2.next = mergeTwoSortedListsRecursive(list1, list2.next);
        return list2;
    }
}

// Helper function to create a linked list from an array
function createLinkedList(values) {
    if (values.length === 0) {
        return null;
    }
    const head = new ListNode(values[0]);
    let current = head;
    for (let i = 1; i < values.length; i++) {
        current.next = new ListNode(values[i]);
        current = current.next;
    }
    return head;
}

// Helper function to convert linked list to array for testing
function linkedListToArray(node) {
    const result = [];
    while (node) {
        result.push(node.val);
        node = node.next;
    }
    return result;
}

// Test cases
let list1 = createLinkedList([1, 2, 4]);
let list2 = createLinkedList([1, 3, 4]);
let result = mergeTwoSortedListsRecursive(list1, list2);
console.log(linkedListToArray(result));  // [1, 1, 2, 3, 4, 4]

list1 = createLinkedList([]);
list2 = createLinkedList([]);
result = mergeTwoSortedListsRecursive(list1, list2);
console.log(linkedListToArray(result));  // []

list1 = createLinkedList([]);
list2 = createLinkedList([0]);
result = mergeTwoSortedListsRecursive(list1, list2);
console.log(linkedListToArray(result));  // [0]

File: solutions/Python/merge_two_sorted_lists_recursive.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def merge_two_sorted_lists_recursive(
    list1: Optional[ListNode], list2: Optional[ListNode]
) -> Optional[ListNode]:
    """
    Merge two sorted linked lists recursively.

    Time Complexity: O(m + n) where m and n are the lengths of list1 and list2
    Space Complexity: O(m + n) due to the recursion call stack
    """
    # Base cases: if one list is empty, return the other
    if not list1:
        return list2
    if not list2:
        return list1

    # Compare the values and recursively merge
    if list1.val <= list2.val:
        list1.next = merge_two_sorted_lists_recursive(list1.next, list2)
        return list1
    else:
        list2.next = merge_two_sorted_lists_recursive(list1, list2.next)
        return list2


# Helper function to create a linked list from a list
def create_linked_list(values):
    if not values:
        return None
    head = ListNode(values[0])
    current = head
    for val in values[1:]:
        current.next = ListNode(val)
        current = current.next
    return head


# Helper function to convert linked list to list for easy testing
def linked_list_to_list(node):
    result = []
    while node:
        result.append(node.val)
        node = node.next
    return result


# Test cases
list1 = create_linked_list([1, 2, 4])
list2 = create_linked_list([1, 3, 4])
result = merge_two_sorted_lists_recursive(list1, list2)
print(linked_list_to_list(result))  # [1, 1, 2, 3, 4, 4]

list1 = create_linked_list([])
list2 = create_linked_list([])
result = merge_two_sorted_lists_recursive(list1, list2)
print(linked_list_to_list(result))  # []

list1 = create_linked_list([])
list2 = create_linked_list([0])
result = merge_two_sorted_lists_recursive(list1, list2)
print(linked_list_to_list(result))  # [0]
